File: api/internal/tailoring/resume_density.go
package tailoring

import (
	"fmt"
	"strings"

	"prismapply/api/internal/profilemode"
)

const (
	minBulletsMostRecentRole = 4
	minBulletsOtherRole      = 2
	minBulletsPerProject     = 2
)

// ResumeDensityHints guides the resume writer toward appropriate length without inventing facts.
type ResumeDensityHints struct {
	PageTarget           string
	ResumeLayout         string
	SeniorityTarget      string
	YearsExperience      string
	MinBulletsMostRecent int
	MinBulletsOtherRole  int
	MinBulletsPerProject int
}

// ExpandFieldHint points the writer at profile fields that can yield additional factual bullets.
type ExpandFieldHint struct {
	Field   string
	Content string
}

func buildResumeDensityHints(profile map[string]any) ResumeDensityHints {
	seniority := profileStr(profile, "seniorityTarget")
	years := profileStr(profile, "yearsExperience")
	layout := profileStr(profile, "resumeLayout")
	if layout == "" {
		layout = profilemode.DeriveResumeLayoutFromDoc(profile)
	}
	page := "1 page"
	if layout == profilemode.LayoutEmploymentLed && isExtendedResumeSeniority(seniority, years) {
		page = "2 pages maximum"
	}

	minRecent := minBulletsMostRecentRole
	minOther := minBulletsOtherRole
	minProject := minBulletsPerProject
	switch layout {
	case profilemode.LayoutProjectOnly:
		minRecent = 0
		minOther = 0
		minProject = 3
	case profilemode.LayoutHybrid:
		minRecent = 2
		minOther = 2
		minProject = 3
	}

	return ResumeDensityHints{
		PageTarget:           page,
		ResumeLayout:         layout,
		SeniorityTarget:      seniority,
		YearsExperience:      years,
		MinBulletsMostRecent: minRecent,
		MinBulletsOtherRole:  minOther,
		MinBulletsPerProject: minProject,
	}
}

func isExtendedResumeSeniority(seniority, years string) bool {
	switch seniority {
	case "staff", "principal", "director", "lead":
		return true
	}
	switch years {
	case "8-12", "12+":
		return true
	}
	return false
}

func collectExpandFieldHints(profile map[string]any) []ExpandFieldHint {
	var hints []ExpandFieldHint
	add := func(field, content string) {
		content = strings.TrimSpace(content)
		if content == "" {
			return
		}
		if len(content) > 1200 {
			content = content[:1200] + "…"
		}
		hints = append(hints, ExpandFieldHint{Field: field, Content: content})
	}

	add("proudestProfessionalWins", profileStr(profile, "proudestProfessionalWins"))
	add("honestCareerNarrative", profileStr(profile, "honestCareerNarrative"))

	storyKeys := []string{
		"storyHardestTechnicalChallenge", "storyDisagreementOrConflict", "storyBiggestMistake",
		"storyLeadingWithoutAuthority", "storyTightDeadline", "storyConflictingPriorities",
		"storyProcessImprovement", "storyDifficultFeedback", "storyMentoringTeaching",
		"storyCrossFunctionalCollaboration", "storyAmbiguousProblem", "storyEthicalOrRiskTradeoff",
	}
	for _, key := range storyKeys {
		add(key, profileStr(profile, key))
	}

	if projects, ok := profile["projects"].([]any); ok {
		for i, pr := range projects {
			pm, ok := pr.(map[string]any)
			if !ok {
				continue
			}
			title := profileStr(pm, "title")
			parts := filterNonEmpty([]string{
				profileStr(pm, "summary"),
				profileStr(pm, "impactMetrics"),
				profileStr(pm, "techStackExtra"),
			})
			if len(parts) == 0 {
				continue
			}
			field := fmt.Sprintf("projects[%d]", i)
			if title != "" {
				field = fmt.Sprintf("projects[%d].%s", i, title)
			}
			add(field, strings.Join(parts, "\n"))
		}
	}

	if entries, ok := profile["workEntries"].([]any); ok {
		for i, item := range entries {
			m, ok := item.(map[string]any)
			if !ok {
				continue
			}
			parts := filterNonEmpty([]string{
				strFrom(m, "role", "Role: "),
				strFrom(m, "company", "Company: "),
				profileStr(m, "summaryBullets"),
			})
			if len(parts) == 0 {
				continue
			}
			add(fmt.Sprintf("workEntries[%d]", i), strings.Join(parts, "\n"))
		}
	}
	return hints
}

func formatDensityHintsBlock(h ResumeDensityHints) string {
	layoutNote := ""
	switch h.ResumeLayout {
	case profilemode.LayoutProjectOnly:
		layoutNote = "Layout: project_only — Skills, Education, Projects only. experience[] must be empty. Do not invent employers."
	case profilemode.LayoutHybrid:
		layoutNote = "Layout: hybrid — Skills, Education, short internship Experience block (from workEntries only), then Projects."
	default:
		layoutNote = "Layout: employment_led — standard Experience then Projects."
	}
	return fmt.Sprintf(`%s
Page target: %s
Applicant seniority: %s
Years experience: %s
Minimum bullets — most recent role: %d; other roles: %d; each included project: %d
Fill the page target using profile-backed detail. Do not pad with generic filler or invented metrics.`,
		layoutNote,
		h.PageTarget,
		orDefault(h.SeniorityTarget, "unspecified"),
		orDefault(h.YearsExperience, "unspecified"),
		h.MinBulletsMostRecent,
		h.MinBulletsOtherRole,
		h.MinBulletsPerProject,
	)
}

func formatExpandHintsBlock(hints []ExpandFieldHint) string {
	if len(hints) == 0 {
		return "(none — use resume and experience sections only)"
	}
	var b strings.Builder
	for _, h := range hints {
		b.WriteString("### ")
		b.WriteString(h.Field)
		b.WriteString("\n")
		b.WriteString(h.Content)
		b.WriteString("\n\n")
	}
	return strings.TrimSpace(b.String())
}

File: orchestration/src/tailoring/resumeDensity.ts
export const MIN_BULLETS_MOST_RECENT_ROLE = 4;
export const MIN_BULLETS_OTHER_ROLE = 2;
export const MIN_BULLETS_PER_PROJECT = 2;

export interface ResumeDensityHints {
  pageTarget: string;
  seniorityTarget: string;
  yearsExperience: string;
  minBulletsMostRecent: number;
  minBulletsOtherRole: number;
  minBulletsPerProject: number;
}

export interface ExpandFieldHint {
  field: string;
  content: string;
}

function profileStr(p: Record<string, unknown>, key: string): string {
  const v = p[key];
  if (typeof v === "string") return v.trim();
  return "";
}

function isExtendedResumeSeniority(seniority: string, years: string): boolean {
  if (["staff", "principal", "director", "lead"].includes(seniority)) return true;
  if (["8-12", "12+"].includes(years)) return true;
  return false;
}

export function buildResumeDensityHints(profile: Record<string, unknown>): ResumeDensityHints {
  const seniorityTarget = profileStr(profile, "seniorityTarget");
  const yearsExperience = profileStr(profile, "yearsExperience");
  const pageTarget = isExtendedResumeSeniority(seniorityTarget, yearsExperience)
    ? "2 pages maximum"
    : "1 page";
  return {
    pageTarget,
    seniorityTarget,
    yearsExperience,
    minBulletsMostRecent: MIN_BULLETS_MOST_RECENT_ROLE,
    minBulletsOtherRole: MIN_BULLETS_OTHER_ROLE,
    minBulletsPerProject: MIN_BULLETS_PER_PROJECT,
  };
}

export function collectExpandFieldHints(profile: Record<string, unknown>): ExpandFieldHint[] {
  const hints: ExpandFieldHint[] = [];
  const add = (field: string, content: string) => {
    const trimmed = content.trim();
    if (!trimmed) return;
    hints.push({
      field,
      content: trimmed.length > 1200 ? `${trimmed.slice(0, 1200)}…` : trimmed,
    });
  };

  add("proudestProfessionalWins", profileStr(profile, "proudestProfessionalWins"));
  add("honestCareerNarrative", profileStr(profile, "honestCareerNarrative"));

  const storyKeys = [
    "storyHardestTechnicalChallenge",
    "storyDisagreementOrConflict",
    "storyBiggestMistake",
    "storyLeadingWithoutAuthority",
    "storyTightDeadline",
    "storyConflictingPriorities",
    "storyProcessImprovement",
    "storyDifficultFeedback",
    "storyMentoringTeaching",
    "storyCrossFunctionalCollaboration",
    "storyAmbiguousProblem",
    "storyEthicalOrRiskTradeoff",
  ];
  for (const key of storyKeys) {
    add(key, profileStr(profile, key));
  }

  const projects = profile.projects;
  if (Array.isArray(projects)) {
    projects.forEach((pr, i) => {
      if (!pr || typeof pr !== "object") return;
      const p = pr as Record<string, unknown>;
      const title = profileStr(p, "title");
      const parts = [
        profileStr(p, "summary"),
        profileStr(p, "impactMetrics"),
        profileStr(p, "techStackExtra"),
      ].filter(Boolean);
      if (parts.length === 0) return;
      let field = `projects[${i}]`;
      if (title) field = `projects[${i}].${title}`;
      add(field, parts.join("\n"));
    });
  }

  return hints;
}

function orDefault(value: string, fallback: string): string {
  return value.trim() ? value : fallback;
}

export function formatDensityHintsBlock(h: ResumeDensityHints): string {
  return `Page target: ${h.pageTarget}
Applicant seniority: ${orDefault(h.seniorityTarget, "unspecified")}
Years experience: ${orDefault(h.yearsExperience, "unspecified")}
Minimum bullets — most recent role: ${h.minBulletsMostRecent}; other roles: ${h.minBulletsOtherRole}; each included project: ${h.minBulletsPerProject}
Fill the page target using profile-backed detail. Do not pad with generic filler or invented metrics.`;
}

export function formatExpandHintsBlock(hints: ExpandFieldHint[]): string {
  if (hints.length === 0) {
    return "(none — use resume and experience sections only)";
  }
  return hints.map((h) => `### ${h.field}\n${h.content}`).join("\n\n");
}
